### DISCUSSION TITLE
workaround  to support <expr> functionality and general plugin development guide.

### DISCUSSION THREAD

**Discussion by sourcevault** [NeoVintageous/NeoVintageous#875] [Q&A]
Continued from here :

⤷ https://github.com/NeoVintageous/NeoVintageous/issues/11#issuecomment-1556136084

⤷ `nnoremap <expr> ' "'" . nr2char(getchar()) . "zz"`

Given its not possible to write something like this natively using the `<expr>` operator what can be a possible workaround ?

I have taken two examples from the issue section that provides hints regarding the process.

Two examples of how to create plugins :

1. https://github.com/NeoVintageous/NeoVintageous/issues/276#issuecomment-334930172

2. https://github.com/NeoVintageous/NeoVintageous/issues/312#issuecomment-372138026

My second question is how do you do effective debugging within `neovintageous` / sublime text plugin ?

My current dev set up is :

1. Use `packageResourceViewer:Extract Package` to get the package file.

2. play around with the code and do debubbing within the sublime text console using :

`
sublime.log_commands(True)
`

Is this a viable method or should I add something else to build something that works ?

**Comment by gerardroche** [comment]
If you are going to work on NeoVintageous I recommend installing manual via a git clone. Then 1 isn't required because you have the source. https://github.com/NeoVintageous/NeoVintageous#manual-installation.

See https://github.com/NeoVintageous/NeoVintageous/blob/master/.github/CONTRIBUTING.md for a guide on enabling debug  logging.

See https://blog.gerardroche.com for some usage of features.

**Comment by gerardroche** [comment]
The only special argument supported in mappings commands is `<silent>` see https://blog.gerardroche.com/2022/11/21/vimrc-and-neovintageousrc/

**Comment by gerardroche** [comment]
I'm not familiar with special arguments like  `<expr>`. What does `nnoremap <expr> ' "'" . nr2char(getchar()) . "zz"` do?

**Reply by sourcevault** [reply]
`<expr> ' "'" . nr2char(getchar()) . "zz"`

 ⤷ It basically says that commands that jumps between marks should follow a `zz` command (move cursor at the middle of the screen ).
 
**AUX ISSUE :** for some reason I am unable to get global marks to work, the ones that is between files and in native vim you use the capital letter to define them.

https://github.com/NeoVintageous/NeoVintageous/assets/20347647/2ee23d00-dad7-474e-924e-39809b7e6361

**Comment by gerardroche** [comment]
Something might help here is to set `scroll_context_lines` in Sublime Text Settings or set `'scrolloff'` which will override the Sublime Text setting. In other words `'scrolloff'` defaults to `scroll_context_lines`.

I set mine to 8 for this very reason.

Menu → Preferences → Settings

```js
"scroll_context_lines": 8,
```

Setting `'scrolloff'` in the neovintageousrc file should work, but I've just noticed a bug. It only sets in on startup for currently opened file and then defaults back to the Sublime Text `scroll_context_lines` setting. Setting `'scrolloff'` on the fly works e.g. `:set scrolloff=8`.


You could map all of them, which is not ideal:

```
nnoremap '' ''zz
nnoremap 'a 'azz
nnoremap 'b 'bzz
" ...
```

I think we could add an optional feature to invoke `zz` after mark jumps. It would be easier that trying to implement all the vimscript functionality.

**Comment by sourcevault** [comment]
> Setting 'scrolloff' in the neovintageousrc file should work

I tried this option too, but it's not ideal. 

> You could map all of them, which is not ideal:

I am going to do this, it's a hard pill to swallow, but it's okay for now. 

I think the problem is that I have really high hope for your wonderful project ( `NeoVintageous` ), in the sense I am hoping it can directly compete with the best vim emulator out there, therefore a non-elegant solution is a hard pill to swallow.

> I think we could add an optional feature to invoke zz after mark jumps. It would be easier that trying to implement all the vimscript functionality.

I think there is no need to introduce something only for this, it sounds like a hackish solution, there are too many edge cases, what if they want `zt` instead of `zz` ? or `zb` ?

A better one would be to let the programmer directly **enable** to write python code to do it.

I do not have knowledge about the internals of your code and it would be too difficult for me to understand it and write something that :

- meets high coding standards.
- works. 
- does not have errors.

What I can try and do, is find a way to write plugin. 

It should be easy to hook into `neovintageous` to monitor user commands to modify it before it execute, or is that not really possible ? 

The final decision regarding all of it is upto you.

**Comment by gerardroche** [comment]
I think implementing things like `nnoremap <expr> ' "'" . nr2char(getchar()) . "zz"` is maybe a bridge too far. I'm not sure how it work.

We essentially want to allow user to map arbitrary keys right? So instead of having to

```
nnoremap 'a 'azz
nnoremap 'b 'bzz
nnoremap 'c 'czz
" ...
```

You can:

```
nnoremap '<char> '<char>zz
```

Which is essentially what `nnoremap <expr> ' "'" . nr2char(getchar()) . "zz"` is doing.

I guess the `<expr>` could be implemented. Initially for basic stuff like `nr2char(getchar())`. But internally it's not easy to resolve things like this. 

`scroll_context_lines` feature in Sublime Text does the job for me, which is why I haven't tried implementing something like `<expr>`. I can't think of an easy way to implement it.

**Reply by sourcevault** [reply]
`scroll_context_lines` is fine if your monitor is set to landscape.

In my case it's a 4k, `16:9` set to vertical 🥴, so that is why I am having such an issue.
